.github/scripts/.build.zsh

```diff
@@ -0,0 +1,347 @@
+          exit 2
+        }
+        target=${2}
+        shift 2
+        ;;
+      -c|--config)
+        if (( ! ${_valid_configs[(Ie)${2}]} )) {
+          log_error "Invalid value %B${2}%b for option %B${1}%b"
+          log_output ${_usage}
+          exit 2
+        }
+        config=${2}
+        shift 2
+        ;;
+      -s|--codesign) codesign=1; shift ;;
+      -q|--quiet) (( verbosity -= 1 )) || true; shift ;;
+      -v|--verbose) (( verbosity += 1 )); shift ;;
+      -h|--help) log_output ${_usage}; exit 0 ;;
+      -V|--version) print -Pr "${_version}"; exit 0 ;;
+      --debug) verbosity=3; shift ;;
+      --generator)
+        if [[ ${host_os} == linux ]] {
+          if (( ! ${_valid_generators[(Ie)${2}]} )) {
+            log_error "Invalid value %B${2}%b for option %B${1}%b"
+            log_output ${_usage}
+            exit 2
+          }
+          generator=${2}
+        }
+        shift 2
+        ;;
+      --print-config) _print_config=1; skips+=(unpack deps); shift ;;
+      --skip-*)
+        local _skip="${${(s:-:)1}[-1]}"
+        local _check=(all deps build)
+        (( ${_check[(Ie)${_skip}]} )) || log_warning "Invalid skip mode %B${_skip}%b supplied"
+        skips+=(${_skip})
+        shift
+        ;;
+      *) log_error "Unknown option: %B${1}%b"; log_output ${_usage}; exit 2 ;;
+    }
+  }
+
+  : "${target:="${host_os}-${CPUTYPE}"}"
+
+  set -- ${(@)args}
+  set_loglevel ${verbosity}
+
+  if (( ! (${skips[(Ie)all]} + ${skips[(Ie)deps]}) )) {
+    check_${host_os}
+  }
+  setup_ccache
+
+  if [[ ${host_os} == linux ]] {
+    autoload -Uz setup_linux && setup_linux
+  }
+
+  local product_name
+  read -r product_name <<< \
+    "$(jq -r '.name' ${buildspec_file})"
+
+  pushd ${project_root}
+  if (( ! (${skips[(Ie)all]} + ${skips[(Ie)build]}) )) {
+    log_group "Configuring ${product_name}..."
+
+    local -a cmake_args=()
+    local -a cmake_build_args=(--build)
+    local -a cmake_install_args=(--install)
+
+    case ${_loglevel} {
+      0) cmake_args+=(-Wno_deprecated -Wno-dev --log-level=ERROR) ;;
+      1) ;;
+      2) cmake_build_args+=(--verbose) ;;
+      *) cmake_args+=(--debug-output) ;;
+    }
+
+    if (( ${+TWITCH_CLIENTID} + ${+TWITCH_HASH} > 1 )) {
+      cmake_args+=(
+        -DTWITCH_CLIENTID:STRING="${TWITCH_CLIENTID}"
+        -DTWITCH_HASH:STRING="${TWITCH_HASH}"
+      )
+    }
+
+    if (( ${+RESTREAM_CLIENTID} + ${+RESTREAM_HASH} > 1 )) {
+      cmake_args+=(
+        -DRESTREAM_CLIENTID:STRING="${RESTREAM_CLIENTID}"
+        -DRESTREAM_HASH:STRING="${RESTREAM_HASH}"
+      )
+    }
+
+    if (( ${+YOUTUBE_CLIENTID} + ${+YOUTUBE_CLIENTID_HASH} + ${+YOUTUBE_SECRET} + ${+YOUTUBE_SECRET_HASH} > 3 )) {
+      cmake_args+=(
+        -DYOUTUBE_CLIENTID:STRING="${YOUTUBE_CLIENTID}"
+        -DYOUTUBE_CLIENTID_HASH:STRING="${YOUTUBE_CLIENTID_HASH}"
+        -DYOUTUBE_SECRET:STRING="${YOUTUBE_SECRET}"
+        -DYOUTUBE_SECRET_HASH:STRING="${YOUTUBE_SECRET_HASH}"
+      )
+    }
```

Comment:
Does this need to be handled here, or can it be handled in a CMake preset?

Reply:
We'd need to generate a preset on the fly by the workflow or action (either using the workflow secrets as input, or an entire preset stored as a base64-encoded string as a secret), but then also adjust the scripts to check for such a preset and use it.

The way the CMake script is set up, the variables should either be set or unset, but using a preset would always set them (albeit with empty values if the secrets are not configured), which would then lead to configuration errors.

Reply:
I don't see how empty strings would break anything. Our CMake explicitly checks for a truthy value (empty strings are FALSE) _and_ that the _HASH variables are set to `(0|[a-fA-F0-9]+)`.

Reply:
IMO it's not worth the complexity of creating a user preset on the fly or editing a user preset template file - setting specific cache variables is a valid approach even when using presets.

Reply:
> If we set cacheVariable entries that use $env or $penv in the CI presets that already exist

We don't right now, because these specific variables are not exposed in the preset (mainly because they are meant for "internal" use of official release builds).

So even if they are set in the environment, the current CMake scripts would ignore them entirely.

I'm not opposed to adding these as default cache variables to the default presets, but that'd be a different PR.

Reply:
I don't understand, but it's not a big deal. It works either way, I'm just confused why this approach is preferred over handling it in presets. Maybe I'll understand it some other day. Either way, it's not worth blocking over.

Reply:
As I mentioned above, the current presets don't contain these variables, so they will also not be adapted from the environment (and this PR was specifically designed not to make any CMake-level changes). And neither does the CMake work with them all being empty:

One of the recent bug fixes enforces the hash being either `0` or a hexadecimal string, but if the environment variable is not set (but used in the CMake preset), it would end up failing that check, so we'd need to add extra code to handle empty hash values (even though they _should_ fail) because they might be passed via the environment, and at this point I'd vote "no" based on complexity.

So my choice was to say "these are user-provided secrets, they don't need to be set to build OBS, and user-provided values arrive either via cache definitions on the command line or via local user presets".

Reply:
As far as I'm concerned, setting these variables to empty strings is equivalent to not setting them at all. That is how the old CMake worked, and that is also how the new CMake works.

Legacy:
https://github.com/obsproject/obs-studio/blob/8d33da1fab53c362d82b9b2fcdcdb02ba4a3a006/UI/cmake/legacy.cmake#L27-L37

New:
https://github.com/obsproject/obs-studio/blob/8d33da1fab53c362d82b9b2fcdcdb02ba4a3a006/UI/cmake/feature-twitch.cmake#L1-L10

I still don't understand how setting them to empty strings would supposedly fail. It works fine in local testing. If empty strings are specified, the feature is (correctly, in my opinion) disabled. Empty strings do not cause a configure or build or runtime failure. No changes in CMakeLists.txt or *.cmake files are required for this. The only required to handle this in a preset is to set the cache variables using env or penv calls, which moves the complexity away from shell scripts.

As mentioned earlier, I won't block on this. I just didn't understand why we were bothering to control it in a script when it can be controlled just fine in a preset, and I thought we were deferring as much config complexity as possible to scripts. Doubly so here, where as far as I know, this script is _meant_ to be run only on CI, and we can create a CI preset that is conditionally only to be enabled on CI if we were so inclined. Or not.